[AutoDiff] Finish #adjoint expression (Sema and SILGen).

dan-zheng opened the pull request:
The purpose of `#adjoint` is to lookup and return the adjoint function
corresponding to a `@differentiable` function.

- Change AdjointExpr to represent original function as a DeclName (and optional
  base type) rather than an expression.
  - The argument of `#adjoint(...)` is a function identifier: this is better
    modeled as a DeclName (and optional base type) rather than an expression,
    which is overly general.
  - Representing the original function as an expression also makes
    type-checking difficult. The relationship between the types of the original
    and adjoint cannot be easily modeled using the constraint system.
- SILGen AdjointExpr.
  - Rather than rewriting AdjointExpr to a different Expr in CSApply (e.g.
    a DeclRefExpr referencing the adjoint function), custom SILGen is used
    because AdjointExpr can represent functions that are normally inaccessible.
    Such DeclRefExprs would be invalid in the AST.

dan-zheng committed:
[AutoDiff] Finish #adjoint expression (Sema and SILGen).

The purpose of #adjoint is to lookup and return the adjoint function
corresponding to a @differentiable function.

- Change AdjointExpr to represent original function as a DeclName (and optional
  base type) rather than an expression.
  - The argument of #adjoint(...) is a function identifier: this is better
    modeled as a DeclName (and optional base type) rather than an expression,
    which is overly general.
  - Representing the original function as an expression also makes
    type-checking difficult. The relationship between the types of the original
    and adjoint cannot be easily modeled using the constraint system.
- SILGen AdjointExpr.
  - Rather than rewriting AdjointExpr to a different Expr in CSApply (e.g.
    a DeclRefExpr referencing the adjoint function), custom SILGen is used
    because AdjointExpr can represent functions that are normally inaccessible.
    Such DeclRefExprs would be invalid in the AST.

dan-zheng committed:
Minor fix for #adjoint expression type-checking test.

Use `self` as an argument for A.B.C.subtract.

dan-zheng committed:
Add #adjoint expression SILGen test.

dan-zheng committed:
Minor test edits.

- Add overloaded original function test.

dan-zheng committed:
Add #adjoint expression test for TensorFlow ops.

- This test is important because #adjoint needs to work with TensorFlow ops to
  unblock ML model creation.

dan-zheng committed:
[NFC] Fix parser grammar comment for `parseFunctionIdentifier`.

rxwei commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -3015,9 +3015,38 @@ visitValueAndGradientExpr(ValueAndGradientExpr *E, SGFContext C) {
+  auto baseMeta = adjointInfo.SILFnType->substGenericArgs(SGF.SGM.M, subs)
+    ->getSelfParameter().getType();
+  auto metatype = SGF.B.createMetatype(loc, SGF.getLoweredType(baseMeta));
+  auto partialApply =
```
Did you mean just "apply"?

dan-zheng replied:
Done, thanks.

rxwei reviewed (commented):
Very nice. Thanks for figuring this out!

dan-zheng committed:
Optimize CSGen for AdjointExpr.

- Shorten custom `addMemberRefConstraints` function: less code duplication.
- Rename `adjointFunc` to `adjointDecl` for consistency.

dan-zheng committed:
Add #adjoint test for matmul.

dan-zheng committed:
Rename `partialApply` to `apply` in SILGenExpr.

Addresses feedback from @rxwei.

dan-zheng commented:
Unfortunately, [`parseFunctionIdentifier`](https://github.com/apple/swift/pull/17324/files#diff-3f6a161822412eea4b0e705487dd88c1R3772) in ParseExpr.cpp duplicates ~150 LOC from [`parseTypeIdentifier`](https://github.com/apple/swift/blob/e307e540985339ff08271aa562cba488edf5c38c/lib/Parse/ParseType.cpp#L548) in ParseType.cpp.

`parseFunctionIdentifier` parses a "function identifier", which consists of an optional base type identifier followed by an unqualified DeclName. It is used for parsing the argument of `#adjoint` expression.

Examples:
```swift
// base type: none, decl name: "+"
#adjoint(+)
// base type: "A.B.C", decl name: "foo"
#adjoint(A.B.C.foo)
// base type: "Tensor<Float>", decl name: "convolved2D"
#adjoint(Tensor<Float>.convolved2D)
```

Most of the function is copied from `parseTypeIdentifier`, which parses the base type. However, custom backtracking logic is necessary to "undo" parsing the final identifier so that it can be parsed as a DeclName.

To avoid code dupe and use `parseTypeIdentifier`, I would need the ability to:
* Backtrack the parser after calling `parseTypeIdentifier` to the position of the last "component"
* "Pop off" the last component in a `IdentTypeRepr`

dan-zheng commented on lib/Sema/TypeCheckExpr.cpp:
```diff
@@ -731,30 +734,35 @@ TypeChecker::lookupFuncDecl(
+  LookupResult results;
+  if (baseType) {
+    results = lookupMember(lookupContext, baseType, funcName);
+  } else {
```
N/A here; we talked in person.

dan-zheng committed:
Clean up `parseFunctionIdentifier`.

dan-zheng commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3756,6 +3756,165 @@ ParserResult<Expr> Parser::parseExprGradientBody(ExprKind kind) {
+/// be kept in sync.
+/// The code is duplicated because custom backtracking logic is needed to parse
+/// the final unqualified decl name.
+static std::tuple<TypeRepr *, DeclName, DeclNameLoc>
```
Done.

dan-zheng commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3756,6 +3756,162 @@ ParserResult<Expr> Parser::parseExprGradientBody(ExprKind kind) {
+///     identifier generic-args? ('.' identifier generic-args?)*
+///
+/// Note: this function is largely copied from `parseTypeIdentifier` and should
+/// be kept in sync.
```
I'll make `parseTypeIdentifier` more general.
Changes to `parseTypeIdentifier` will be a bit invasive, but I think it's worth it because the level of dupe is not acceptable and `parseQualifiedDeclName` is a generally useful function.

I'll name the flag `isParsingQualifiedDeclName` for now: open to name suggestions.

dan-zheng committed:
[Parse] Add `isParsingQualifiedDeclName` flag to `parseTypeIdentifier`.

- Rename `parseFunctionIdentifier` to `parseQualifiedDeclName`.
- Add flag to `parseTypeIdentifier` so that `parseQualifiedDeclName` can use it.
- Use `parseQualifiedDeclName` for #adjoint expression.

The changes to `parseTypeIdentifier` were slightly invasive, but the change is
worthwhile to reduce code duplication. `parseQualifiedDeclName` is also a
generally useful function.

Also, fix @differentiable type checking errors (don't pass in base type to
`TC.lookupFuncDecl` for better diagnostic).

rxwei reviewed (approved):
Nice. Thanks!

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

mhong commented on test/AutoDiff/adjoint_expr_silgen.swift:
```diff
@@ -0,0 +1,210 @@
+}
+// CHECK-LABEL: sil hidden @test_top_level
+// CHECK: [[DSQUARE_RAW:%.*]] = function_ref @dsquare : $@convention(thin) <τ_0_0 where τ_0_0 : FloatingPoint> (@in_guaranteed τ_0_0, @in_guaranteed τ_0_0, @in_guaranteed τ_0_0) -> @out τ_0_0
+// CHECK: [[DSQUARE:%.*]] = partial_apply [callee_guaranteed] [[DSQUARE_RAW]]<Float>
```
omit "[[DSQUARE:%.*]] = " to keep the test code simple?

same below

dan-zheng replied:
I would rather keep it for clarity, if you don't mind.

mhong replied:
How does this help with clarify? I feel this is akin to defining a variable but not using it.

dan-zheng replied:
It is the case that a variable is declared without being used, but the name of the variable provides clarity.

dan-zheng replied:
I removed the unused variables in https://github.com/apple/swift/pull/17324/commits/4f13c8f533f8a990b1bda37ac2ad5e64810b2a9b. Thanks for the feedback!

mhong commented on test/AutoDiff/adjoint_expr_silgen.swift:
```diff
@@ -0,0 +1,210 @@
+  return adjointSquare(1, 2, 3)
+}
+// CHECK-LABEL: sil hidden @test_top_level
+// CHECK: [[DSQUARE_RAW:%.*]] = function_ref @dsquare : $@convention(thin) <τ_0_0 where τ_0_0 : FloatingPoint> (@in_guaranteed τ_0_0, @in_guaranteed τ_0_0, @in_guaranteed τ_0_0) -> @out τ_0_0
```
do we care about the details of " (@in_guaranteed τ_0_0, @in_guaranteed τ_0_0, @in_guaranteed τ_0_0) -> @out τ_0_0" in this test? if not, it'd be good to omit them, to avoid unnecessary test maintenance work in the future.

e.g. in case @in_guaranteed is updated to some other calling convention.

dan-zheng replied:
I agree that the extra test maintenance is unpleasant, but I would like to keep the function signatures in the test because they're informative.

Many function_refs are (partially) applied and it's difficult to tell what's happening without seeing the function type. For instance methods, it's especially important to verify that `#adjoint` has the intended type (it should be curried, first taking a `self` instance argument).

If the default calling convention changes frequently, I'd be happy to refactor this test.

mhong replied:
how does the SIL code here help verify that it takes a self instance argument? it seems that info will need to be in the next line: on partial_apply

mhong replied:
How do we read the function_ref inst to determine it's being curried or not?

dan-zheng replied:
Here's an toy example adapted from `adjoint_expr_silgen.swift`:
```swift
// Original Swift.
extension Pair {
  @differentiable(reverse, ...)
  func subtract(_ a: Pair) -> Pair { ... }

  private func dSubtract(_ a: Pair, primal: Pair, seed: Pair) -> (Pair, Pair) { ... }
}

@_silgen_name("test_pair")
func testPair() {
  _ = #adjoint(Pair.subtract)
}
```
```llvm
// Output from `swiftc -emit-silgen`.
// test_pair
sil hidden @test_pair : $@convention(thin) () -> () {
bb0:
  // function_ref curry thunk of Pair.dSubtract(_:primal:seed:)
  %0 = function_ref @$xxx : $@convention(thin) (Pair) -> @owned @callee_guaranteed (Pair, Pair, Pair) -> (Pair, Pair) // user: %1
  ... // omitted
} // end sil function 'test_pair'
```

The type of the method `xxx` is (ignoring calling convention markers) `(Pair) -> (Pair, Pair, Pair) -> (Pair, Pair)`. This is currying, like the example you provided above.

Normally in Swift, when applying an instance method like `x.dSubtract(...)`, the `self` instance (in this case `x`) is applied implicitly.
However, if you reference an instance method directly using `Pair.dSubtract`, its type will be curried (because there's no `self` instance).

For `#adjoint` expression, we decided to make `#adjoint(<instance method>)` always have a curried type because that's more general and flexible.

mhong replied:
Thanks! This is very helpful.

mhong commented on lib/AST/ASTDumper.cpp:
```diff
@@ -1834,9 +1834,17 @@ class PrintExpr : public ExprVisitor<PrintExpr> {
-    OS << ')';
+    PrintWithColorRAII(OS, TypeReprColor) << " base_type='";
+    if (auto baseRepr = E->getBaseType().getTypeRepr())
+      baseRepr->print(PrintWithColorRAII(OS, TypeReprColor).getOS());
```
write: "auto *baseRepr" for consistency with existing code

dan-zheng replied:
Done.

mhong commented on lib/Parse/ParseType.cpp:
```diff
@@ -545,38 +545,54 @@ bool Parser::parseGenericArguments(SmallVectorImpl<TypeRepr*> &Args,
-      return makeParserCodeCompletionResult<IdentTypeRepr>();
-    }
+/// SWIFT_ENABLE_TENSORFLOW: Added `isParsingQualifiedDeclName` flag.
+/// The `isParsingQualifiedDeclName` flag controls whether parsing is done as if
```
this comment block seems to better belong to the header / interface.

dan-zheng replied:
Done.

mhong commented on lib/Parse/ParseType.cpp:
```diff
@@ -545,38 +545,54 @@ bool Parser::parseGenericArguments(SmallVectorImpl<TypeRepr*> &Args,
-    }
+/// SWIFT_ENABLE_TENSORFLOW: Added `isParsingQualifiedDeclName` flag.
+/// The `isParsingQualifiedDeclName` flag controls whether parsing is done as if
+/// this type identifier is the prefix to a qualified declaration name. If true,
```
can you explain what "qualified declaration name" is, and/or give an example?

also, why do we need to add this knob -- what's special about the AD expr type compared to other expr types?

(the change to this func seems pretty invasive and is more likely to cause merge conflicts than other s4tf changes. so it'd be useful to spend extra care and assess if there are better alternatives.)

mhong replied:
Thanks for the pointers! I did read the PR description. I'll try and read other pre-existing PR comments in future reviews as well (somtimes reviewers may not be reading all review exchanges with other parties).

Name overloading is not special to AD. Usually a language has precedence rules to help decide which one to choose, like what's being implemented here -- essentially when there are multiple matches, pick the one with the scope closest to the use-site. 

I'm guess there is an pre-existing impl in swift (is it in type checker, for ranking candidates and picking the best fit one?). it'd be great to try and find that and reuse. please file a bug to track this.

dan-zheng replied:
I mentioned overloaded names as the reason for parsing a qualified name rather than an unqualified one.
However, the actual overload resolution isn't handled here - that's done in CSGen via an existing function [`TypeChecker::lookupFuncDecl`](https://github.com/apple/swift/blob/4316cf9f25cb441ef0108a703874dc85058c60e1/lib/Sema/CSGen.cpp#L1468).
Here, the base type and function name are parsed. These are passed to `lookupFuncDecl` as arguments.

rxwei replied:
Thanks for weighing in and asking questions! I think the examples you listed ("name overloading", "precedence rules", etc) are largely irrelevant to the topic being discussed here. This particular code piece is adding backtracking logic to enable code sharing and parsing a fully specified decl name, not related to AD or overload resolution. "Qualified decl name" has a well-defined meaning throughout Swift's AST, and represents a complete lookup path. And there is no pre-existing parser logic to parse a qualified decl name **without** parsing an expression.

In this particular case, the code author has implementation experience and established knowledge about how parser and type checking works, and the implementation a) does generally follow the established convention and b) does have clear explanation of each decision made. Before drawing speculations based largely on assumptions, I would appreciate it if you read more about the how Parse and Sema work.

mhong replied:
From a follow-up discussion with Dan, I captured these points:

- Before this change, there is no concept/support of a qualified flavor of a decl name in Swift compiler. A func name like `A.B` would be an expression (DeclRefExpr), and name overloading is solved with that mechanism.
- But for AD infra, using an expression for the adjoint func name is evaluated and decided against. Hence we are extending decl name to support the qualified flavor.

Thanks for the additional context and clarification!

rxwei replied:
Great! We'll just have to take extra care to incorporate this upstream.

mhong commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -3015,9 +3015,37 @@ visitValueAndGradientExpr(ValueAndGradientExpr *E, SGFContext C) {
+  SILDeclRef adjointDeclRef(adjointDecl);
+
+  // If adjoint has multiple parameter lists, mark as curried.
+  if (adjointDecl->getNumParameterLists() == 2) {
```
>= 2?

this will also better match the comment above

dan-zheng replied:
Done. Existing code uses `getNumParameterLists() == 2`, however.

mhong replied:
I mean ">= 2" -- the formatter ate ">" in my previous message :)

mhong reviewed (changes requested):
Nice! Posted some comments and questions (I'm still learning about the AD infra).

dan-zheng committed:
Minor edits.

Address comments from @mhong.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please clean test tensorflow

dan-zheng commented:
@swift-ci Please clean test tensorflow

dan-zheng commented:
@swift-ci Please clean test tensorflow

dan-zheng committed:
Remove unused FileCheck variables.

shahmishal commented:
@swift-ci Please clean test tensorflow

shahmishal commented:
@swift-ci Please clean test tensorflow

dan-zheng commented:
Thanks for reviewing!
